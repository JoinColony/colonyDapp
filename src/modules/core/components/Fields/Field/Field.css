@import 'variables.css';

.main {
  line-height: 1;
}

.main + .main {
  margin-top: 20px;
}

.input {
  width: 100%;
  border: 1px solid var(--temp-grey-5);
  border-radius: var(--radius-tiny);
  font-size: var(--size-normal);
  color: var(--text);
  letter-spacing: var(--spacing-normal);
  transition: border-color 0.1s ease-in;

  &:focus {
    border-color: var(--primary);
  }

  &::placeholder {
    font-size: var(--size-normal);
    font-style: italic;
    letter-spacing: var(--spacing-normal);
  }
}

.main[aria-invalid="true"] .input {
  border-color: var(--danger);
}

/* States override all existing styles */
.stateActive .input {
  border-color: var(--primary) !important;
}

.displayInline {
  display: inline-block;
}

.checkDelegate {
  height: 16px;
  width: 16px;
  position: absolute;
  visibility: hidden;
}

.error, .help {
  display: inline-block;
  margin-left: 5px;
  font-size: var(--size-tiny);
  font-weight: 400;
  line-height: 1;
}

.error {
  color: var(--danger);
}

.help {
  color: var(--text);
}

.errorHorizontal {
  composes: error;
  display: block;
  margin-top: 5px;
  height: 1.2em;
  position: absolute;
  top: 100%;
  right: 0;
  text-align: right;
}
