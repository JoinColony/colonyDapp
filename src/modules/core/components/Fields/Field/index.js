/* @flow */

import type { ComponentType } from 'react';
import { compose } from 'recompose';
import withProps from 'recompose/withProps';
import { injectIntl } from 'react-intl';

import FieldComponent from './Field.jsx';

export { default } from './Field.jsx';

export const asField = (props?: Object) => (component: ComponentType<Object>) => compose(
  injectIntl,
  withProps((ownProps?: Object) => ({
    component,
    ...ownProps,
    ...props,
  })),
)(FieldComponent);

// exports a field without using redux-form
// it will not be connected to redux-form so take care of validation and stuff yourself
export const asStandaloneField = (props?: Object) => asField({
  ...props,
  standalone: true,
});
