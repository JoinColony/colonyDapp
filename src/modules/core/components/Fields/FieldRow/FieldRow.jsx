/* @flow */

import React, { Component } from 'react';
import type { ComponentType } from 'react';
import type { FieldProps } from 'redux-form';
import type { Appearance } from '$types/css';

import type { FieldComponentProps, Option, Utils } from '../flowTypes';

type Props = FieldProps & {
  appearance?: Appearance,
  elementOnly?: boolean,
  fieldComponent: ComponentType<FieldComponentProps<>>,
  help: string,
  id: string,
  immediateValidation?: boolean,
  label: string,
  name: string,
  options?: Array<Option>,
  placeholder: string,
  utils: Utils,
  title: string,
};

class FieldRow extends Component<Props> {
  reset: () => void;
  static displayName = 'core.Field.FieldRow';
  constructor(props: Props) {
    super(props);
    this.reset = this.reset.bind(this);
  }
  reset() {
    const { input: { onChange } } = this.props;
    onChange('');
  }
  render() {
    const {
      appearance,
      custom,
      elementOnly,
      fieldComponent: FieldComponent,
      help,
      id,
      immediateValidation,
      input,
      label,
      meta,
      options,
      placeholder,
      utils,
      title,
      ...props
    } = this.props;

    const { touched, pristine, error } = meta;
    const hasError = !!(error && (touched || (immediateValidation && !pristine)));

    const inputProps = {
      id,
      placeholder,
      title: hasError ? (error || title) : title,
    };

    return (
      <FieldComponent
        appearance={appearance}
        custom={custom}
        elementOnly={elementOnly}
        error={error}
        hasError={hasError}
        help={help}
        input={input}
        inputProps={inputProps}
        label={label}
        meta={meta}
        options={options}
        passthroughProps={props}
        utils={{
          ...utils,
          reset: utils.reset || this.reset,
        }}
      />
    );
  }
}
export default FieldRow;
