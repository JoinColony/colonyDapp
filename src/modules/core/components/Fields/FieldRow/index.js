/* @flow */

export { default } from './FieldRow.jsx';
