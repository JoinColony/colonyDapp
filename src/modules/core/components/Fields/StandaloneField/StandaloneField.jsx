/* @flow */

import React, { Component } from 'react';
import type { ComponentType } from 'react';

type Props = {
  component: ComponentType<*>,
};

type State = {
  active: boolean,
};

class StandaloneField extends Component<Props, State> {
  setActive: () => void;
  setInactive: () => void;
  constructor(props: Props) {
    super(props);
    this.state = {
      active: false,
    };
    this.setActive = this.setActive.bind(this);
    this.setInactive = this.setInactive.bind(this);
  }
  setActive() {
    this.setState({
      active: true,
    });
  }
  setInactive() {
    this.setState({
      active: false,
    });
  }
  render() {
    const { component: RowComponent, ...props } = this.props;
    const { active } = this.state;
    return (
      <RowComponent {...props} onFocus={this.setActive} onBlur={this.setInactive} meta={{ active }} />
    );
  }
}

export default StandaloneField;
