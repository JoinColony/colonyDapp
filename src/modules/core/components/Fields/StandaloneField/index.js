/* @flow */

export { default } from './StandaloneField.jsx';
