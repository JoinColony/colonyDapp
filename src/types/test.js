/* @flow */

import {
  Address,
  ENSName,
  type AddressType,
  type ENSNameType,
} from './strings';

const takesENSName = (arg: ENSNameType) => arg;
const takesAddress = (arg: AddressType) => arg;

const name = ENSName('ens name');
const address = Address('0x123');

takesENSName(name);
takesAddress(address);

/*
 * The following should produce flow errors:
 */
// Address('not a hex string');
// takesAddress('0x123'); // just a string
// takesENSName('just a string');
