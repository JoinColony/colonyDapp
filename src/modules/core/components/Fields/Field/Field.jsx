/* @flow */

import React, { Component } from 'react';
import type { ComponentType } from 'react';
import PropTypes from 'prop-types';
import { Field as ReduxFormField } from 'redux-form';
import type { IntlShape, MessageDescriptor } from 'react-intl';
import shortid from 'shortid';

import type { Appearance } from '$types/css';

import type { FieldComponentProps, IntlFormatter, Option } from '../flowTypes';

import FieldRow from '../FieldRow';
import StandaloneField from '../StandaloneField';

/**
 * Component to easily create Field components that are connected to redux-form
 *
 * @method Field

 * Components wrapped as a Field *get* the following props:
 *
 * @param {object} appearance Object to set the Fields appearance
 * @param {bool} elementOnly Set to just display the `input` element w/o it's label
 * @param {any} error The error message to display in case of non-valid values, can be set to a messageDescriptor
 * @param {bool} hasError Used to determine if to the error message is to be shown
 * @param {string|object} help The help message to display, can be set to a messageDescriptor (optional)
 * @param {string} input Holds all the properties that the actual input element should get (think: redux-forms input but enhanced)
 * @param {string|object} label The label of the `input` field, can be set to a messageDescriptor
 * @param {array} options Options for radio groups
 * @param {func} reset Reset method provided by `redux-form`.
 */

type Props = {
  appearance?: Appearance,
  component: ComponentType<FieldComponentProps<>>,
  elementOnly?: boolean,
  help?: MessageDescriptor | string,
  id?: string,
  intl: IntlShape,
  label?: MessageDescriptor | string,
  name: string,
  options?: Array<Option>,
  placeholder?: MessageDescriptor | string,
  standalone?: boolean,
  title?: MessageDescriptor | string,
  reset?: () => void,
};

type Context = {
  _reduxForm: Object,
};

class Field extends Component<Props> {
  getLabel: () => string;
  getIntlFormatted: IntlFormatter;
  getTitle: () => string;
  getConnectorComponent: () => ComponentType<*>;
  id: string;
  static contextTypes = {
    _reduxForm: PropTypes.object,
  };
  constructor(props: Props, context: Context) {
    super(props, context);
    this.getLabel = this.getLabel.bind(this);
    this.getIntlFormatted = this.getIntlFormatted.bind(this);
    this.getTitle = this.getTitle.bind(this);
    this.getConnectorComponent = this.getConnectorComponent.bind(this);
  }
  componentWillMount() {
    const { _reduxForm } = this.context;
    const { elementOnly, id, label, name } = this.props;
    if (!_reduxForm) {
      this.id = `${name}_${shortid.generate()}`;
    } else {
      this.id = `${_reduxForm.form}_${name}_${shortid.generate()}`;
    }
    if (elementOnly && !label && !id) {
      console.error('WARNING: You have to specify an id when using an external label');
    }
  }
  getId() {
    return this.props.id || this.id;
  }
  getIntlFormatted(prop?: MessageDescriptor | string, values?: { [string]: string }): string {
    const { intl: { formatMessage } } = this.props;
    if (!prop) {
      return '';
    }
    if (typeof prop == 'string') {
      return prop;
    }
    return formatMessage(prop, values);
  }
  getLabel(): string {
    const { label } = this.props;
    return this.getIntlFormatted(label);
  }
  getTitle() {
    const { label, placeholder, title } = this.props;
    return this.getIntlFormatted(title) || this.getIntlFormatted(label) || this.getIntlFormatted(placeholder);
  }
  getConnectorComponent(): (ComponentType<*> | Function | string) {
    const { standalone } = this.props;
    return standalone ? StandaloneField : ReduxFormField;
  }
  render() {
    const {
      component,
      help,
      intl,
      options,
      placeholder,
      reset,
      standalone,
      ...props
    } = this.props;

    const ConnectorComponent = this.getConnectorComponent();
    const connectorProps = {
      ...props,
      fieldComponent: component,
      help: this.getIntlFormatted(help),
      id: this.getId(),
      label: this.getLabel(),
      options: options && options.map(option => ({ ...option, label: this.getIntlFormatted(option.label) })),
      placeholder: this.getIntlFormatted(placeholder),
      utils: {
        getIntlFormatted: this.getIntlFormatted,
        reset,
      },
      title: this.getTitle(),
    };

    return (
      <ConnectorComponent
        component={FieldRow}
        {...connectorProps}
      />
    );
  }
}

export default Field;
