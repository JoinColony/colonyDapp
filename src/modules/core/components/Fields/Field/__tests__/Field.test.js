/* eslint-env jest */

import React from 'react';
import PropTypes from 'prop-types';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Field as ReduxFormField } from 'redux-form';
import { shallowWithIntl, mountWithIntl } from '$utils/test';

import Field from '../Field.jsx';
import StandaloneField from '../../StandaloneField';
import FieldRow from '../../FieldRow';

jest.mock('shortid', () => ({
  generate: jest.fn(() => 'foo'),
}));

jest.mock('../../FieldRow/FieldRow.jsx');

const TestFieldComponent = () => <span />;

const middlewares = [];
const mockStore = configureStore(middlewares);

describe('Field component standalone', () => {
  test('Renders initial component', () => {
    const wrapper = shallowWithIntl((
      <Field name="test" standalone component={TestFieldComponent} />
    ));
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(StandaloneField)).toBePresent();
  });

  test('Receives actual fieldComponent as prop', () => {
    const wrapper = shallowWithIntl((
      <Field name="test" standalone component={TestFieldComponent} />
    ));
    expect(wrapper).toHaveProp('fieldComponent', TestFieldComponent);
  });

  test('Assembles a proper id for the field', () => {
    const wrapper = shallowWithIntl((
      <Field name="test" standalone component={TestFieldComponent} />
    ));
    expect(wrapper).toHaveProp('id', 'test_foo');
  });

  test('Uses the id provided if specified', () => {
    const wrapper = shallowWithIntl((
      <Field name="test" id="cool" standalone component={TestFieldComponent} />
    ));
    expect(wrapper).toHaveProp('id', 'cool');
  });

  test('Passes props to field wrapper', () => {
    const wrapper = shallowWithIntl((
      <Field
        className="className"
        name="test"
        data-foo="bar"
        standalone
        component={TestFieldComponent}
      />
    ));
    expect(wrapper.find(StandaloneField)).toHaveProp('className', 'className');
    expect(wrapper).toHaveProp('data-foo', 'bar');
  });

  test('Warns when using it without a rendered label and without an id', () => {
    const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
    shallowWithIntl((
      <Field name="test" standalone component={TestFieldComponent} elementOnly />
    ));
    expect(consoleSpy).toHaveBeenCalled();
  });

  test('Assembles labels properly', () => {
    const wrapperStringLabel = shallowWithIntl((
      <Field name="test" elementOnly label="Test" standalone component={TestFieldComponent} />
    ));
    expect(wrapperStringLabel).toHaveProp('label', 'Test');
    const wrapperMsgLabel = shallowWithIntl((
      <Field name="test" elementOnly label={{ id: 'foo', defaultMessage: 'Label' }} standalone component={TestFieldComponent} />
    ));
    expect(wrapperMsgLabel).toHaveProp('label', 'Label');
  });
});

describe('Field component embedded in redux-form', () => {
  const reduxFormStub = {
    form: 'stub',
    register: jest.fn(),
    getFormState: jest.fn(),
  };
  const reduxFormContext = {
    context: { _reduxForm: reduxFormStub, store: mockStore({}) },
    childContextTypes: { _reduxForm: PropTypes.object, store: PropTypes.object },
  };
  test('Contains a ReduxFormField', () => {
    const wrapper = mountWithIntl(
      (
        <Field name="test" label="Test" component={TestFieldComponent} />
      ),
      reduxFormContext,
    );
    expect(wrapper.find(ReduxFormField)).toBePresent();
  });
  test('Assembles field id', () => {
    const wrapper = mountWithIntl(
      (
        <Field name="test" label="Test" component={TestFieldComponent} />
      ), reduxFormContext,
    );
    expect(wrapper.find(FieldRow)).toHaveProp('id', 'stub_test_foo');
  });
});
