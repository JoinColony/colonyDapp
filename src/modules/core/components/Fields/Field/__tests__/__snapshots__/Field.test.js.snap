// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component standalone Renders initial component 1`] = `
<StandaloneField
  component={[Function]}
  fieldComponent={[Function]}
  help=""
  id="test_foo"
  label=""
  name="test"
  placeholder=""
  title=""
  utils={
    Object {
      "getIntlFormatted": [Function],
      "reset": undefined,
    }
  }
/>
`;
